tidy3d/components/data_new.py

```diff
@@ -0,0 +1,137 @@
+import json
+
+import pydantic as pd
+import h5py
+import xarray as xr
+
+from .base import Tidy3dBaseModel
+from .types import ArrayLike
+from .simulation import Simulation
+from .monitor import FieldMonitor, FieldTimeMonitor, FluxMonitor, FluxTimeMonitor, ModeMonitor
+from .monitor import PermittivityMonitor, ModeFieldMonitor
+
+# every type that can be used as a component
+ComponentType = Union[
+    Simulation,
+    FieldMonitor,
+    FieldTimeMonitor,
+    FluxMonitor,
+    FluxTimeMonitor,
+    ModeMonitor,
+    ModeFieldMonitor,
+]
+
+
+class Tidy3dData(Tidy3dBaseModel):
+
+    component: ComponentType = pd.Field(
+        None,
+        title="Original component",
+        description="Copy of the tidy3d component that this data is associated with.",
+    )
+
+    data: xr.DataArray = pd.Field(
+        None,
+        title="Data",
+        description="An xarray DataArray corresponding to the data stord by this element.",
+    )
+
+    data_dict: Dict[str, "Tidy3dData"] = pd.Field(
+        {},
+        title="Data dictionary",
+        description="Collection of other tidy3d data stored by this element",
+    )
+
+    @property
+    def coords(self) -> Dict[str, ArrayLike]:
+        """Labelled coordinates corresponding to the axes of ``self.values``."""
+        return self.data.coords
+
+    @property
+    def values(self) -> ArrayLike:
+        """The raw data associated with the data, represented as a multi-dimensional array."""
+        return self.data.values
+
+    @property
+    def dims(self) -> Tuple[str]:
+        """Tuple of the ordered coordinate labels corresponding to each axis of ``self.values.``"""
+        return self.data.dims
+
+    def __getitem__(self, name) -> "Tidy3dData":
+        """Return the ``Tidy3dData`` associated with the collection member using square brackets."""
+        item = self.data_dict.get(name)
+        if item is None:
+            raise Tidy3dKeyError(f'item "{item}" not found in data_dict.')
+        return item
+
+    @property
+    def json_string(self) -> str:
+        """A string representation of everything execpt for the ``.data`` and ``.data_dict``."""
+        return self.json(exclude={"data", "data_dict"})
+
+    """ Saving to an hdf5 file."""
+
+    def save_component(self, f_handle, hdf5_grp):
+        """Convert everything but the ``.data`` and ``.data_dict`` to a string and save to group."""
+        hdf5_grp.create_dataset("json", data=self.json_string)
+
+    def save_data(self, f_handle, hdf5_grp):
+        """Save the ``xarray.DataArray`` to an hdf5 group."""
+        hdf5_grp.create_dataset("values", data=self.values)
+        for dim in self.dims:
+            coord = self.coords[dim]
+            hdf5_grp.create_dataset(dim, data=coord)
+
+    def save_data_dict(self, f_handle, hdf5_grp):
+        """Save the ``data_dict`` to an hdf5 group recursively."""
+        for name, tidy3d_data in self.data_dict.items():
+            new_group = f_handle.create_group(name)
```

Comment:
It should be `hdf5_grp.create_group` I believe. You should be able to use group handles as file handles and `f_handle` should not be needed.

Reply:
ok I just saw this, will check that.  would be better for sure